README
######################################################################
    Net::SSH::AuthorizedKeysFile 0.17
######################################################################

NAME
    Net::SSH::AuthorizedKeysFile - Read and modify ssh's authorized_keys
    files

SYNOPSIS
        use Net::SSH::AuthorizedKeysFile;

            # Reads $HOME/.ssh/authorized_keys by default
        my $akf = Net::SSH::AuthorizedKeysFile->new();

        $akf->read("authorized_keys");

            # Iterate over entries
        for my $key ($akf->keys()) {
            print $key->as_string(), "\n";
        }

            # Modify entries:
        for my $key ($akf->keys()) {
            $key->option("from", 'quack@quack.com');
            $key->keylen(1025);
        }
            # Save changes back to $HOME/.ssh/authorized_keys
        $akf->save() or die "Cannot save";

DESCRIPTION
    Net::SSH::AuthorizedKeysFile reads and modifies "authorized_keys" files.
    "authorized_keys" files contain public keys and meta information to be
    used by "ssh" on the remote host to let users in without having to type
    their password.

METHODS
    "new"
        Creates a new Net::SSH::AuthorizedKeysFile object and reads in the
        authorized_keys file. The filename defaults to
        "$HOME/.ssh/authorized_keys" unless overridden with

            Net::SSH::AuthorizedKeysFile->new( file => "/path/other_authkeys_file" );

        Normally, the "read" method described below will just silently
        ignore faulty lines and only gobble up keys that either one of the
        two parsers accepts. If you want it to be stricter, set

            Net::SSH::AuthorizedKeysFile->new( file   => "authkeys_file",
                                               abort_on_error => 1 );

        and read will immediately abort after the first faulty line. Also,
        the key parsers are fairly lenient in default mode. Adding

            strict => 1

        adds sanity checks before a key is accepted.

    "read"
        Reads in the file defined by new(). By default, strict mode is off
        and read() will silently ignore faulty lines. If it's on (see new()
        above), read() will immediately abort after the first faulty line. A
        textual description of the last error will be available via error().

    "content"
        Contains the original file content, read by "read()" earlier. Can be
        used to set arbitrary content:

            $keysfile->content( "some\nrandom\nlines\n" );

        and have "parse()" operate on a string instead of an actual file
        this way.

    "keys"
        Returns a list of Net::SSH::AuthorizedKey objects. Methods are
        described in Net::SSH::AuthorizedKey.

    "as_string"
        String representation of all keys, ultimately the content that gets
        written out when calling the "save()" method. Note that comments
        from the original file are lost.

    "save"
        Write changes back to the authorized_keys file using the as_string()
        method described above. Note that comments from the original file
        are lost. Optionally takes a file name parameter, so calling
        "$akf->save("foo.txt")" will save the data in the file "foo.txt"
        instead of the file the data was read from originally. Returns 1 if
        successful, and undef on error. In case of an error, error()
        contains a textual error description.

    "sanity_check"
        Run a sanity check on the currently selected authorized_keys file.
        If it contains insanely long lines, then parsing with read() (and
        potential crashes because of out-of-memory errors) should be
        avoided.

    "ssh_dir( [$user] )"
        Locate the .ssh dir of a given user. If no user name is given,
        ssh_dir will look up the .ssh dir of the effective user. Typically
        returns something like "/home/gonzo/.ssh".

    "path_locate( [$user] )"
        Locate the authorized_keys file of a given user. Typically returns
        something like "/home/gonzo/.ssh/authorized_keys". See "ssh_dir()"
        for how the containing directory is located with and without a given
        user name.

    "error"
        Description of last error that occurred.

LEGALESE
    Copyright 2005-2009 by Mike Schilli, all rights reserved. This program
    is free software, you can redistribute it and/or modify it under the
    same terms as Perl itself.

AUTHOR
    2005, Mike Schilli <m@perlmeister.com>



Changes
######################################################################
Revision history for Perl extension Net::SSH::AuthorizedKeysFile
######################################################################

0.17  (12/22/2014)
    (ms) David Leon Gil added support for Ed25519 keys

0.16  (01/31/2012)
    (ms) Added support for ecdsa-* ssh-2 keys, requested by
         Christian Ruppert.

0.15  (02/16/2011)
    (ms) No longer using $& to eliminate both real and imaginary
         performance penalties.

0.14  (12/20/2010)
    (ms) Preserve order of options
    (ms) Bug fix: Binary options no longer expand as xxx="1" in as_string()

0.13  (2010/04/25)
    (ms) Moved ssh1 and ssh2 docs to base class.
    (ms) Added sanity check for files with extreme line lengths
    (ms) Added convenience methods path_locate() and ssh_dir() to locate 
         a user's .ssh dir and authorized_keys file on disk.
    (ms) Added fingerprint method to parser base class, determining
         a key's functional uniqueness.

0.12  (2010/02/22)
    (ms) Added test cases with weird keys
    (ms) Deleting undefined/empty options to avoid warnings

0.11  (2009/11/30)
    (ms) Documentation fixes
    (ms) Added github repo link to Makefile.PL

0.10  2009/11/15
    (ms) Complete rewrite of parsers and classes
    (ms) No longer supporting multi-line keys

0.06  2009/10/20
    (ms) new() no longer reads the configuration file
         automatically, read() must be called separately. This breaks
         backward compatibility. Sorry, but it had to be fixed.
  
    (ms) Even odd cases of authorized_keys files are now handled 
         correctly (but of course there could be even odder cases
         that caused the parser to bail :)

0.05  2009/08/07
    (ms) Ignore empty lines
    (ms) Allow multi-lines for ssh2 pubkeys
    (ms) Fixes for more resilience towards mistyped authkey files

0.04  2009/08/07
    (ms) Allow mixed v1/v2 keys in a single authorized_keys file.

0.03  2008/08/24
    (ms) Rewrote entire parsing code to comply with ssh spec, thanks to 
         Tobias Galitzien for the pointer.
    (ms) Ignore comment lines in authorized_keys files
    (ms) The 'email' field is now call 'comment' ('email' is still available 
         for backwards compatibility, though).
    (ms) Comments can now contain blanks and commas without tripping the
         parser up.

0.02  2005/12/23
    (ms) Fixed dependencies

0.01  2005/12/21
    (ms) Where it all began.


Makefile.PL
######################################################################
# Makefile.PL for Net::SSH::AuthorizedKeysFile
# 2005, Mike Schilli <m@perlmeister.com>
######################################################################
use ExtUtils::MakeMaker;

my $meta_merge = {
    META_MERGE => {
        resources => {
            repository  => 'http://github.com/mschilli/net-ssh-authorizedkeysfile-perl',
        },
    }
};

WriteMakefile(
    'NAME'         => 'Net::SSH::AuthorizedKeysFile',
    'VERSION_FROM' => 'lib/Net/SSH/AuthorizedKeysFile.pm', # finds $VERSION
    'PREREQ_PM'    => {
        Test::More       => 0,
        Text::ParseWords => 0,
        Log::Log4perl    => 1,
        File::Temp       => 0,
        File::Copy       => 0,
    }, # e.g., Module::Name => 1.1
    $ExtUtils::MakeMaker::VERSION >= 6.50 ? (%$meta_merge) : (),
    'EXE_FILES'    => [ 'eg/authorized-keys-test' ],
    ($] >= 5.005 ?    ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM => 'lib/Net/SSH/AuthorizedKeysFile.pm',
       AUTHOR     => 'Mike Schilli <m@perlmeister.com>') : ()),
);


lib/Net/SSH/AuthorizedKey.pm
###########################################
package Net::SSH::AuthorizedKey;
###########################################
use strict;
use warnings;
use Log::Log4perl qw(:easy);

use Net::SSH::AuthorizedKey::SSH1;
use Net::SSH::AuthorizedKey::SSH2;

###########################################
sub parse {
###########################################
    my($class, $string) = @_;

    my @subclasses = qw(
        Net::SSH::AuthorizedKey::SSH1
        Net::SSH::AuthorizedKey::SSH2
    );

    for my $subclass ( @subclasses ) {
        DEBUG "Parsing with $subclass: $string";
        my $pk = $subclass->parse( $string );
        if($pk) {
            DEBUG "Successfully parsed $subclass key";
            return $pk;
        }
    }

    return undef;
}

1;

__END__

=head1 NAME

Net::SSH::AuthorizedKey - Virtual Base Class for SSH Public Keys

=head1 SYNOPSIS

    use Net::SSH::AuthorizedKey;

      # Either parse a string (without leading whitespace or comments):
    my $key = Net::SSH::AuthorizedKey->parse( $line );

    if(defined $key) {
          # ssh-1 or ssh-2
        print "Key parsed, type is ", $key->type(), "\n";
    } else {
        die "Cannot parse key '$line'";
    }

      # ... or create an object yourself:
    my $pubkey = Net::SSH::AuthorizedKey->new(
        options  => { from                  => 'foo@bar.com', 
                      "no-agent-forwarding" => 1 },
        key      => "123....890",
        keylen   => 1024,
        exponent => 35,
        type     => "ssh-1",
    );

=head1 DESCRIPTION

Net::SSH::AuthorizedKey is a virtual base class for ssh public keys. 
Real implementations of it are Net::SSH::AuthorizedKey::SSH1 and
Net::SSH::AuthorizedKey::SSH2.

The only way to using it directly is by calling its parse() method, and passing
it an authorized_keys string (aka a line from an authorized_keys file). If it
recognizes either a ssh-1 or a ssh-2 type key, it will return a
Net::SSH::AuthorizedKey::SSH1 or a Net::SSH::AuthorizedKey::SSH2 object, both
of which support the accessor methods defined in the FIELDS section below.

The as_string() method will cobble the (perhaps modified) fields together
and return them as a string suitable as a line for an authorized_keys file.

=head2 METHODS

=over 4

=item C<parse( $line )>

Reads in a single text line containing a ssh-1 or ssh-2 key.
Returns a Net::SSH::AuthorizedKey::SSH1 or a Net::SSH::AuthorizedKey::SSH2
object, or C<undef> in case of an error.

=item C<error()>

Returns the last parsing error encountered as a text string.

=item C<as_string( )>

Return the object as a string suitable as a autorized_keys line.

=back

=head2 FIELDS

All of the following fields are available via accessors:

=over 4

=item C<type>

Type of ssh key, usually C<"ssh-1"> or C<"ssh-2">.

=item C<key>

Public key, either a long number (ssh-1) or a line of alphanumeric
characters (ssh-2).

=item C<keylen>

Length of the key in bit (e.g. 1024).

=item C<exponent>

Two-digit number in front of the key in ssh-1 authorized_keys lines.

=item C<options>

Returns a reference to a hash with options key/value pairs, listed in 
front of the key.

=back

=head2 IMPLEMENTATION REFERENCE

The key parsers implemented in this distribution are implemented similarily
as the authorized_keys file parser in the openssh source distribution.

Openssh contains the authorized_keys parser
in its auth2_pubkey.c file. The user_key_allowed2() function opens
the file and reads it line by line, ignoring leading whitespace, empty
and comment lines.

After that, if a line doesn't contain a plain key, the parser skips ahead until
the first whitespace (zooming through quoted areas "..." and interpreting '\"'
as an escaped quote), then skips this whitespace and tries to read a key one
more time. 

Regarding options, the Perl parser isn't as elaborate with semantic 
peculiarities as openssh's auth_parse_options(), but this might be
added in the future.

=head1 NOTES FOR SUBCLASS DEVELOPERS

If you're just using Net::SSH::AuthorizedKey to parse keys, the
following section doesn't concern you. It's only relevant if you add 
new subclasses to this package, on top of what's already provided.

Net::SSH::AuthorizedKey is a (semi-)virtual base class that implements 
options handling for its SSH1 and SSH2 subclasses.

SSH key lines can contain options that carry values (like command="ls") and
binary options that are either set or unset (like "no_agent_forwarding"). To
distinguish the two, and to provide a set of allowed option names, the subclass
has to implement the method option_type(), which takes an option name, and
returns

=over 4

=item *

undef if the option is not supported

=item *

"s" if the option is a "string" option that carries a value

=item *

1 if the option is a binary option

=back

The subclasses Net::SSH::AuthorizedKey::SSH1 and Net::SSH::AuthorizedKey::SSH2
are doing this already.

=head1 LEGALESE

Copyright 2005-2009 by Mike Schilli, all rights reserved.
This program is free software, you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 AUTHOR

2005, Mike Schilli <m@perlmeister.com>


lib/Net/SSH/AuthorizedKey/Base.pm
###########################################
package Net::SSH::AuthorizedKey::Base;
###########################################
use strict;
use warnings;
use Log::Log4perl qw(:easy);
use Text::ParseWords;
use Digest::MD5 qw(md5_hex);

  # Accessors common for both ssh1 and ssh2 keys
our @accessors = qw(key type error email comment);
__PACKAGE__->make_accessor( $_ ) for @accessors;

  # Some functions must be implemented in the subclass
do {
    no strict qw(refs);

    *{__PACKAGE__ . "::$_"} = sub {
        die "Whoa! '$_' in the virtual base class has to be ",
            " implemented by a real subclass.";
    };

} for qw(option_type as_string);

  # Options accepted by all keys
our %VALID_OPTIONS = (
    "no-port-forwarding"  => 1,
    "no-agent-forwarding" => 1,
    "no-x11-forwarding"   => 1,
    "no-pty"              => 1,
    "no-user-rc"          => 1,
    command               => "s",
    environment           => "s",
    from                  => "s",
    permitopen            => "s",
    tunnel                => "s",
);

###########################################
sub new {
###########################################
    my($class, %options) = @_;

    my $self = {
        error        => "(no error)",
        option_order => [],
        %options,
    };

    bless $self, $class;
    return $self;
}

###########################################
sub option_type_global {
###########################################
    my($self, $key) = @_;

    if(exists $VALID_OPTIONS{ $key }) {
        return $VALID_OPTIONS{ $key };
    }

      # Maybe the subclass knows about it
    return $self->option_type($key);
}

###########################################
sub options {
###########################################
    my($self) = @_;

    return {
        map { $_ => $self->option( $_ ) } 
            keys %{ $self->{ options } } 
    };
}

###########################################
sub option {
###########################################
    my($self, $key, $value, $append) = @_;

    $key = lc $key;

    my $option_type = $self->option_type_global($key);

    if(! defined $option_type) {
        LOGWARN "Illegal option '$key'";
        return undef;
    }

    if(defined $value) {

        if( $append ) {
            if( $self->{options}->{$key} and
                ref($self->{options}->{$key}) ne "ARRAY" ) {
                $self->{options}->{$key} = [ $self->{options}->{$key} ];
            }
        } else {
            $self->option_delete( $key );
        }

        if($option_type eq "s") {
            if( $self->{options}->{$key} and
                ref($self->{options}->{$key}) eq "ARRAY" ) {
                DEBUG "Adding option $key to $value";
                push @{ $self->{options}->{$key} }, $value;
            } else {
                DEBUG "Setting option $key to $value";
                $self->{options}->{$key} = $value;
            }
        } else {
            $self->{options}->{$key} = undef;
        }
        push @{ $self->{option_order} }, $key;
    }

    if( "$option_type" eq "1" ) {
        return exists $self->{options}->{$key};
    }

    return $self->{options}->{$key};
}

###########################################
sub option_delete {
###########################################
    my($self, $key) = @_;

    $key = lc $key;

    @{ $self->{option_order} } = 
        grep { $_ ne $key } @{ $self->{option_order} };

    delete $self->{options}->{$key};
}

###########################################
sub options_as_string {
###########################################
    my($self) = @_;

    my $string = "";
    my @parts  = ();

    for my $option ( @{ $self->{option_order} } ) {
        if(defined $self->{options}->{$option}) {
            if(ref($self->{options}->{$option}) eq "ARRAY") {
                for (@{ $self->{options}->{$option} }) {
                    push @parts, option_quote($option, $_);
                }
            } else {
                push @parts, option_quote($option, $self->{options}->{$option});
            }
        } else {
            push @parts, $option;
        }
    }
    return join(',', @parts);
}

###########################################
sub option_quote {
###########################################
    my($option, $text) = @_;

    $text =~ s/([\\"])/\\$1/g;
    return "$option=\"" . $text . "\"";
}

###########################################
sub parse {
###########################################
    my($class, $string) = @_;

    DEBUG "Parsing line '$string'";

    # Clean up leading whitespace
    $string =~ s/^\s+//;
    $string =~ s/^#.*//;
 
    if(! length $string) {
        DEBUG "Nothing to parse";
        return;
    }

    if(my $key = $class->key_read( $string ) ) {
          # We found a key without options
        $key->{options} = {};
        DEBUG "Found ", $key->type(), " key: ", $key->as_string();
        return $key;
    }

    # No key found. Probably there are options in front of the key.
    # By the way: the openssh-5.x parser doesn't allow escaped 
    # backslashes (\\), so we don't either.
    (my $key_string = $string) =~ s/^(
                                  (?:"(?:\\"|.)*?)"|
                                  \S
                                )+
                               //x;
    my $options_string =  $&;
    $key_string        =~ s/^\s+//;

    DEBUG "Trying line with options stripped: [$key_string]";

    if(my $key = $class->key_read( $key_string ) ) {
          # We found a key with options
        $key->{options} = {};
        $key->options_parse( $options_string );
        DEBUG "Found ", $key->type(), " key: ", $key->as_string();
        return $key;
    }

    DEBUG "$class cannot parse line: $string";

    return undef;
}

###########################################
sub options_parse {
###########################################
    my($self, $string) = @_;

    DEBUG "Parsing options: [$string]";
    my @options = parse_line(qr/\s*,\s*/, 0, $string);

      # delete empty/undefined fields
    @options = grep { defined $_ and length $_ } @options;

    DEBUG "Parsed options: ", join(' ', map { "[$_]" } @options);

    for my $option (@options) {
        my($key, $value) = split /=/, $option, 2;
        $value = 1 unless defined $value;
        $value =~ s/^"(.*)"$/$1/; # remove quotes

        $self->option($key, $value, 1);
    }
}

###########################################
sub fingerprint {
###########################################
    my($self) = @_;

    my $data = $self->options();

    my $string = join '', map { $_ => $data->{$_} } sort keys %$data;
    $string .= $self->key();

    return md5_hex($string);
}

##################################################
# Poor man's Class::Struct
##################################################
sub make_accessor {
##################################################
    my($package, $name) = @_;

    no strict qw(refs);

    my $code = <<EOT;
        *{"$package\\::$name"} = sub {
            my(\$self, \$value) = \@_;

            if(defined \$value) {
                \$self->{$name} = \$value;
            }
            if(exists \$self->{$name}) {
                return (\$self->{$name});
            } else {
                return "";
            }
        }
EOT
    if(! defined *{"$package\::$name"}) {
        eval $code or die "$@";
    }
}

1;

__END__

=head1 NAME

Net::SSH::AuthorizedKey::Base - Virtual Base Class for ssh keys

=head1 SYNOPSIS

    # Documentation to understand methods shared
    # by all parsers. Not for direct use.

=head1 DESCRIPTION

This is the key parser base class, offering methods common to all
parsers. Don't use it directly, but read the documentation below to
see what functionality all parsers offer.

=over 4

=item error()

If a parser fails for any reason, it will leave a textual description of
the error that threw it off. This methods retrieves the error text.

=item options()

=item key()

The actual content of the key, either a big number in case of ssh-1 or
a base64-encoded string for ssh-2.

=item type()

Type of a key. (Somewhat redundant, as you could also check what subclass
a key is of). Either set to C<"ssh-1"> or C<"ssh-2">.

=item email()

Identical with comment().

=item comment()

Identical with email(). This is the text that follows in the authorized_keys
file after the key content. Mostly used for emails and host names.

=back

=head1 LEGALESE

Copyright 2005-2009 by Mike Schilli, all rights reserved.
This program is free software, you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 AUTHOR

2005, Mike Schilli <m@perlmeister.com>


lib/Net/SSH/AuthorizedKeysFile.pm
###########################################
package Net::SSH::AuthorizedKeysFile;
###########################################
use strict;
use warnings;
use Log::Log4perl qw(:easy);
use Text::ParseWords;
use Net::SSH::AuthorizedKey;
use Net::SSH::AuthorizedKey::SSH1;
use Net::SSH::AuthorizedKey::SSH2;

our $VERSION = "0.17";

###########################################
sub new {
###########################################
    my($class, @options) = @_;

    my $self = {
        default_file        => "$ENV{HOME}/.ssh/authorized_keys",
        strict              => 0,
        abort_on_error      => 0,
        append              => 0,
        ridiculous_line_len => 100_000,
        @options,
    };

    bless $self, $class;

      # We allow keys to be set in the constructor
    my $keys = $self->{keys} if exists $self->{keys};

    $self->reset();

    $self->{keys} = $keys if defined $keys;

    return $self;
}

###########################################
sub sanity_check {
###########################################
    my($self, $file) = @_;

    $self->{file} = $file if defined $file;
    $self->{file} = $self->{default_file} if !defined $self->{file};

    my $result = undef;

    my $fh;

    if(! open $fh, "<$self->{file}") {
        ERROR "Cannot open file $self->{file}";
        return undef;
    }

    while(
      defined(my $rc = 
              sysread($fh, my $chunk, $self->{ridiculous_line_len}))) {
        if($rc < $self->{ridiculous_line_len}) {
            $result = 1;
            last;
        }

        if(index( $chunk, "\n" ) >= 0) {
              # contains a newline, looks good
            next;
        }

          # we've got a line that's between ridiculous_line_len and
          # 2*ridiculous_line_len characters long. Pull the plug.
        $self->error("File $self->{file} contains insanely long lines " .
                     "(> $self->{ridiculous_line_len} chars");
        last;
    }

DONE:
    close $fh;

    if(!$result) {
        ERROR "Sanity check of file $self->{file} failed";
    }
    return $result;
}

###########################################
sub keys {
###########################################
    my($self) = @_;

    return @{$self->{keys}};
}

###########################################
sub reset {
###########################################
    my($self) = @_;

    $self->{keys}    = [];
    $self->{content} = "";
    $self->{error}   = undef;
}

###########################################
sub content {
###########################################
    my($self, $new_content) = @_;

    if( defined $new_content ) {
       $self->reset();
       $self->{content} = $new_content;
    }

    return $self->{content};
}

###########################################
sub line_parse {
###########################################
    my($self, $line, $line_number) = @_;

    chomp $line;

    DEBUG "Parsing line [$line]";

    $self->error( "" );

    my $pk = Net::SSH::AuthorizedKey->parse( $line );

    if( !$pk ) {
        my $msg = "[$line] rejected by all parsers";
        WARN $msg;
        $self->error($msg);
        return undef;
    }

    if(! $self->{strict} or $pk->sanity_check()) {
        return $pk;
    }

    WARN "Key [$line] failed sanity check";

    if($self->{strict}) {
        $self->error( $pk->error() );
        return undef;
    }

      # Key is corrupted, but ok in non-strict mode
    return $pk;
}

###########################################
sub parse {
###########################################
    my($self) = @_;

    $self->{keys}  = [];
    $self->{error} = "";

    my $line_number = 0;

    for my $line (split /\n/, $self->{content}) {
        $line_number++;

        $line =~ s/^\s+//;     # Remove leading blanks
        $line =~ s/\s+$//;     # Remove trailing blanks
        next if $line =~ /^$/; # Ignore empty lines
        next if $line =~ /^#/; # Ignore comment lines

        my $key = $self->line_parse($line, $line_number);

        if( defined $key ) {
            push @{$self->{keys}}, $key;
        } else {
            if($self->{abort_on_error}) {
                $self->error("Line $line_number: " . $self->error());
                return undef;
            }
        }
    }

    return 1;
}

###########################################
sub read {
###########################################
    my($self, $file) = @_;

    $self->reset();

    $self->{file} = $file if defined $file;
    $self->{file} = $self->{default_file} if !defined $self->{file};  
    $self->{content} = "";

    DEBUG "Reading in $self->{file}";

    open FILE, "<$self->{file}" or LOGDIE "Cannot open $self->{file}";

    while(<FILE>) {
        $self->{content} .= $_;
    }

    close FILE;

   return $self->parse();
}

###########################################
sub as_string {
###########################################
    my($self) = @_;

    my $string = "";

    for my $key ( @{ $self->{keys} } ) {
        $string .= $key->as_string . "\n";
    }

    return $string;
}

###########################################
sub save {
###########################################
    my($self, $file) = @_;

    if(!defined $file) {
        $file = $self->{file};
    }

    if(! open FILE, ">$file") {
        $self->error("Cannot open $file ($!)");
        WARN $self->error();
        return undef;
    }

    print FILE $self->as_string();
    close FILE;
}

###########################################
sub append {
###########################################
    my($self, $key) = @_;

    $self->{append} = 1;
}

###########################################
sub error {
###########################################
    my($self, $text) = @_;

    if(defined $text) {
        $self->{error} = $text;
        ERROR "$text";
    }

    return $self->{error};
}

###########################################
sub ssh_dir {
###########################################
    my($self, $user) = @_;

    if(!defined $user) {
        my $uid = $>;
        $user = getpwuid($uid);
        if(!defined $user) {
            ERROR "getpwuid of $uid failed ($!)";
            return undef;
        }
    }

    my @pwent = getpwnam($user);

    if(! defined $pwent[0]) {
        ERROR "getpwnam of $user failed ($!)";
        return undef;
    }

    my $home = $pwent[7];

    return File::Spec->catfile($home, ".ssh");
}

###########################################
sub path_locate {
###########################################
    my($self, $user) = @_;

    my $ssh_dir = $self->ssh_dir($user);

    return undef if !defined $ssh_dir;

    return File::Spec->catfile($ssh_dir, "authorized_keys");
}

1;

__END__

=head1 NAME

Net::SSH::AuthorizedKeysFile - Read and modify ssh's authorized_keys files

=head1 SYNOPSIS

    use Net::SSH::AuthorizedKeysFile;

        # Reads $HOME/.ssh/authorized_keys by default
    my $akf = Net::SSH::AuthorizedKeysFile->new();

    $akf->read("authorized_keys");

        # Iterate over entries
    for my $key ($akf->keys()) {
        print $key->as_string(), "\n";
    }

        # Modify entries:
    for my $key ($akf->keys()) {
        $key->option("from", 'quack@quack.com');
        $key->keylen(1025);
    }
        # Save changes back to $HOME/.ssh/authorized_keys
    $akf->save() or die "Cannot save";

=head1 DESCRIPTION

Net::SSH::AuthorizedKeysFile reads and modifies C<authorized_keys> files.
C<authorized_keys> files contain public keys and meta information to
be used by C<ssh> on the remote host to let users in without 
having to type their password.

=head1 METHODS

=over 4

=item C<new>

Creates a new Net::SSH::AuthorizedKeysFile object and reads in the 
authorized_keys file. The filename 
defaults to C<$HOME/.ssh/authorized_keys> unless
overridden with

    Net::SSH::AuthorizedKeysFile->new( file => "/path/other_authkeys_file" );

Normally, the C<read> method described below will just silently ignore 
faulty lines and only gobble up keys that either one of the two parsers
accepts. If you want it to be stricter, set

    Net::SSH::AuthorizedKeysFile->new( file   => "authkeys_file",
                                       abort_on_error => 1 );

and read will immediately abort after the first faulty line. Also, 
the key parsers are fairly lenient in default mode. Adding

    strict => 1

adds sanity checks before a key is accepted.

=item C<read>

Reads in the file defined by new(). By default, strict mode is off and 
read() will silently ignore faulty lines. If it's on (see new() above),
read() will immediately abort after the first faulty line. A textual
description of the last error will be available via error().

=item C<content>

Contains the original file content, read by C<read()> earlier. Can be
used to set arbitrary content:

    $keysfile->content( "some\nrandom\nlines\n" );

and have C<parse()> operate on a string instead of an actual file 
this way.

=item C<keys>

Returns a list of Net::SSH::AuthorizedKey objects. Methods are described in
L<Net::SSH::AuthorizedKey>.

=item C<as_string>

String representation of all keys, ultimately the content that gets
written out when calling the C<save()> method. 
Note that comments from the original file are lost.

=item C<save>

Write changes back to the authorized_keys file using the as_string()
method described above. Note that comments from the original file are lost.
Optionally takes a file
name parameter, so calling C<$akf-E<gt>save("foo.txt")> will save the data
in the file "foo.txt" instead of the file the data was read from originally.
Returns 1 if successful, and undef on error. In case of an error, error()
contains a textual error description.

=item C<sanity_check>

Run a sanity check on the currently selected authorized_keys file. If
it contains insanely long lines, then parsing with read() (and potential
crashes because of out-of-memory errors) should be avoided.

=item C<ssh_dir( [$user] )>

Locate the .ssh dir of a given user. If no user name is given, ssh_dir will
look up the .ssh dir of the effective user. Typically returns something like
"/home/gonzo/.ssh".

=item C<path_locate( [$user] )>

Locate the authorized_keys file of a given user. Typically returns something 
like "/home/gonzo/.ssh/authorized_keys". See C<ssh_dir()> for how the 
containing directory is located with and without a given user name.

=item C<error>

Description of last error that occurred.

=back

=head1 LEGALESE

Copyright 2005-2009 by Mike Schilli, all rights reserved.
This program is free software, you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 AUTHOR

2005, Mike Schilli <m@perlmeister.com>
